### PR TITLE
[翻译完毕] write test

### LINKED ISSUE #108: [翻译完毕] write_test

### PR DESCRIPTION
**Pull request by isdanni**
Thought we could continue the translation work :)

Plus all the docs can still be useful once migrated to taichi.graphics. Thanks!

# issue

close #108

### CODE REVIEW

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 1]
```diff
@@ -3,61 +3,70 @@
...
 
 #: ../../write_test.rst:6
 msgid ""
-"We use `pytest <https://github.com/pytest-dev/pytest>`_ for our Python "
-"test infrastructure."
+"We use `pytest <https://github.com/pytest-dev/pytest>`_ for our Python test "
+"infrastructure."
 msgstr ""
+"我们将 `pytest <https://github.com/pytest-dev/pytest>`_ 用于我们的Python测试基础"
+"架构。"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"我们将 `pytest <https://github.com/pytest-dev/pytest>`_ 作为 Python 测试的基础"
"架构。"
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 2]
```diff
@@ -3,61 +3,70 @@
...
-"test. If not, feel free to create a new file for it :) So in this case "
-"let's create a new file ``tests/python/test_logarithm.py`` for "
-"simplicity."
+"Look into ``tests/python``, see if there's already a file suit for your test. "
+"If not, feel free to create a new file for it :) So in this case let's create a "
+"new file ``tests/python/test_logarithm.py`` for simplicity."
 msgstr ""
+"在 ``tests/python`` 中查看是否已经有适合你测试的文件。 如果没有，请随时为它创建"
+"一个新文件 :) 因此，在这种情况下，为简单起见，让我们创建一个新文件 ``tests/"
+"python/test_logarithm.py`` 。"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"在 ``tests/python`` 中查看是否已经有适合你测试的文件。 如果没有，请随时为它创建"
"一个新文件 :) 如此，按照上述示例让我们创建一个新文件 ``tests/"
"python/test_logarithm.py`` 即可。"
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 3]
```diff
@@ -66,13 +75,19 @@ msgid ""
...
 msgid ""
-"Add some simple code make use of our ``ti.log10`` to make sure it works "
-"well. Hint: You may pass/return values to/from Taichi-scope using 0-D "
-"tensors, i.e. ``r[None]``."
+"Add some simple code make use of our ``ti.log10`` to make sure it works well. "
+"Hint: You may pass/return values to/from Taichi-scope using 0-D tensors, i.e. "
+"``r[None]``."
 msgstr ""
+"使用我们的 ``ti.log10`` 添加一些简单的代码，以确保其正常工作。 提示：你可以使用 "
+"0-D 张量（即 ``r[None]`` ）向/从Taichi-scope传递值或从其中返回值。"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"利用我们的 ``ti.log10`` 添加一些简单代码，以确保其正常工作。 提示：你可以使用 "
"0-D 张量（即 ``r[None]`` ）向 Taichi-scop 传递值或从中返回。"
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 4]
```diff
@@ -91,25 +106,44 @@ msgid ""
...
 
 #: ../../write_test.rst:48
 msgid ""
-"Execute ``ti test logarithm``, and the functions starting with ``test_`` "
-"in ``tests/python/test_logarithm.py`` will be executed."
+"Execute ``ti test logarithm``, and the functions starting with ``test_`` in "
+"``tests/python/test_logarithm.py`` will be executed."
 msgstr ""
+"执行 ``ti test logarithm`` ，然后将执行 ``tests/python/test_logarithm.py`` 中以 "
+"``test_`` 开头的函数。"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"执行 ``ti test logarithm`` ，随后 ``tests/python/test_logarithm.py`` 中以 "
"``test_`` 开头的函数就会被运行。"
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 5]
```diff
@@ -91,25 +106,44 @@ msgid ""
...
-"we provide a useful decorator ``@ti.test``:"
+"The above method is not good enough, for example, ``ti.init(arch=ti.cpu)``, "
+"means that it will only test on the CPU backend. So do we have to write many "
+"tests ``test_log10_cpu``, ``test_log10_cuda``, ... with only the first line "
+"different? No worries, we provide a useful decorator ``@ti.test``:"
 msgstr ""
+"上面的方法还不够好，例如， ``ti.init(arch=ti.cpu)`` ，意味着它将仅在CPU后端进行"
+"测试。 那么，我们是否必须编写许多测试 ``test_log10_cpu`` ，"
+"``test_log10_cuda`` ，...，而只有第一行不同？ 不用担心，我们提供了一个有用的装饰"
+"器 ``@ti.test`` ："
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"上面的方法还不够好，例如， ``ti.init(arch=ti.cpu)`` ，意味着它将仅在CPU后端进行"
"测试。 那么，我们是否必须编写繁多的测试文件，诸如： ``test_log10_cpu`` ，"
"``test_log10_cuda`` ，...，而它们只有第一行的代码是不同的？ 不用担心，我们提供了一个有用的装饰"
"器 ``@ti.test`` ："
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 6]
```diff
@@ -152,27 +199,45 @@ msgid ""
...
 
 #: ../../write_test.rst:93
 msgid "Cool! Right? But that's still not good enough."
-msgstr ""
+msgstr "很酷不是吗？ 但是，这还不够好。"
 
 #: ../../write_test.rst:96
 msgid "Using ``ti.approx`` for comparison with tolerance"
-msgstr ""
+msgstr "使用 ``ti.approx`` 与公差进行比较"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po:223]:
```suggestion
msgstr "使用 ``ti.approx`` 提高比较操作时的容忍度"
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 7]
```diff
@@ -191,39 +256,58 @@ msgid ""
...
 msgid ""
-"Simply using ``pytest.approx`` won't work well here, since it's tolerance"
-" won't vary among different Taichi backends. It'll be likely to fail on "
-"the OpenGL backend."
+"Simply using ``pytest.approx`` won't work well here, since it's tolerance won't "
+"vary among different Taichi backends. It'll be likely to fail on the OpenGL "
+"backend."
 msgstr ""
+"直接使用 ``pytest.approx`` 在这里表现不是很好，因为它的容忍度在不同的 Taichi 后"
+"端之间不会有所不同。 而它在OpenGL后端上可能会失败。"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"直接使用 ``pytest.approx`` 在这里表现不是很好，因为它的容忍度在不同的 Taichi 后"
"端间差别不是很大。 如此在 OpenGL 后端上它可能会失效。"
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 8]
```diff
@@ -191,39 +256,58 @@ msgid ""
...
 
 #: ../../write_test.rst:133
 msgid ""
-"For example, ``r[None] = 100``, means that it will only test the case of "
-"``ti.log10(100)``. What if ``ti.log10(10)``? ``ti.log10(1)``?"
+"For example, ``r[None] = 100``, means that it will only test the case of ``ti."
+"log10(100)``. What if ``ti.log10(10)``? ``ti.log10(1)``?"
 msgstr ""
+"例如，``r[None] = 100`` ，意味着它将仅测试 ``ti.log10(100)`` 的情况。 那如果 "
+"``ti.log10(10)`` 怎么办？ ``ti.log10(1)`` 呢？"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"例如，``r[None] = 100`` ，意味着它将仅仅测试 ``ti.log10(100)`` 的情况。 那如果 "
"``ti.log10(10)`` 怎么办？ ``ti.log10(1)`` 呢？"
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 9]
```diff
@@ -244,10 +328,26 @@ msgid ""
...
+"        r[None] = ti.log10(r[None])\n"
+"\n"
+"    r[None] = x\n"
+"    foo()\n"
+"    assert r[None] == math.log10(x)"
 
 #: ../../write_test.rst:157
 msgid "Use a comma-separated list for multiple input values:"
-msgstr ""
+msgstr "将逗号分隔的表用于多个输入值："
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po:350]:
```suggestion
msgstr "对于多个输入值，使用逗号分割列表："
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 10]
```diff
@@ -270,12 +370,30 @@ msgid ""
...
+"    s[None] = y\n"
+"    foo()\n"
+"    assert r[None] == math.atan2(x, y)"
 
 #: ../../write_test.rst:180
 msgid ""
 "Use two separate ``parametrize`` to test **all combinations** of input "
 "arguments:"
-msgstr ""
+msgstr "使用两个单独的 ``参数'`` 来测试输入参数的 **所有组合**："
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
msgstr "使用两个独立的 ``参数'`` 来测试输入参数的 **所有组合**："
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 11]
```diff
@@ -300,64 +418,97 @@ msgid ""
...
 "    # ... (some tests have to be done in debug mode)"
 msgstr ""
+"def test_debugging_utils():\n"
+"    ti.init(arch=ti.cpu, debug=True, log_level=ti.TRACE)\n"
+"    # ...（某些测试必须在调试模式下完成）"
 
 #: ../../write_test.rst:225
 msgid "Exclude some backends from test"
-msgstr ""
+msgstr "从测试中排除一些后端"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
msgstr "在测试时排除一些后端"
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 12]
```diff
@@ -300,64 +418,97 @@ msgid ""
...
 "def test_sparse_tensor():\n"
-"    # ... (some tests that requires sparse feature which is not supported"
-" by OpenGL)"
+"    # ... (some tests that requires sparse feature which is not supported by "
+"OpenGL)"
 msgstr ""
+"# 在除OpenGL之外的所有后端上运行此测试 \n"
+"@ti.test(excludes=[ti.opengl])\n"
+"def test_sparse_tensor():\n"
+"    # ...（某些测试需要OpenGL不支持的稀疏特征）"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"    # ...（某些测试需要稀疏特征，而这是 OpenGL 所不支持的）"
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 13]
```diff
@@ -300,64 +418,97 @@ msgid ""
...
 
 #: ../../write_test.rst:236
 msgid ""
 "You may also use the ``extensions`` keyword to exclude backends without "
 "specific feature:"
-msgstr ""
+msgstr "你也可以使用 ``扩展`` 关键字来排除没有特定功能的后端："
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
msgstr "你也可以使用 ``extensions`` 关键字来排除没有特定功能的后端："
```

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 14]
```diff
@@ -300,64 +418,97 @@ msgid ""
...
-"    # ... (some tests that requires sparse feature which is not supported"
-" by OpenGL)"
+"    # ... (some tests that requires sparse feature which is not supported by "
+"OpenGL)"
 msgstr ""
-
+"# 在除OpenGL之外的所有后端上运行此测试 \n"
+"@ti.test(extensions=[ti.extension.sparse])\n"
+"def test_sparse_tensor():\n"
+"    # ...（某些测试需要OpenGL不支持的稀疏特征"
```
Comment (ArkhamWJZ) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"    # ...（某些测试需要稀疏特征，而这是 OpenGL 所不支持的)"
```

**Review by ArkhamWJZ — COMMENTED** [review]
LGTM！Only a few minor details need to be changed 🌟

File: locales/zh_CN/LC_MESSAGES/write_test.po [hunk 15]
```diff
@@ -191,39 +256,58 @@ msgid ""
...
 #: ../../write_test.rst:123
 msgid ""
-"Simply using ``pytest.approx`` won't work well here, since it's tolerance"
-" won't vary among different Taichi backends. It'll be likely to fail on "
-"the OpenGL backend."
+"Simply using ``pytest.approx`` won't work well here, since it's tolerance won't "
+"vary among different Taichi backends. It'll be likely to fail on the OpenGL "
+"backend."
 msgstr ""
+"直接使用 ``pytest.approx`` 在这里表现不是很好，因为它的容忍度在不同的 Taichi 后"
```
Comment (isdanni) [inline locales/zh_CN/LC_MESSAGES/write_test.po]:
```suggestion
"直接使用 ``pytest.approx`` 在这里表现不是很好，因为它的公差在不同的 Taichi 后"
```

My bad. `Tolerance` here should be a mathematics term:

https://www.mathsisfun.com/definitions/tolerance.html

**Review by ArkhamWJZ — APPROVED** [review]

**Pull request merged**